// @progress/kendo-react-inputs/rating/RatingItem.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as n from "react";
import a from "prop-types";
import { useDir as m, classNames as y } from "@progress/kendo-react-common";
const c = n.forwardRef((e, i) => {
  const l = n.useRef(null), o = n.useRef(null), u = m(o, e.dir);
  n.useImperativeHandle(l, () => ({
    element: o.current,
    props: e,
    value: e.value
  })), n.useImperativeHandle(i, () => l.current);
  const s = n.useCallback(
    (t) => {
      e.onClick && e.onClick.call(
        void 0,
        {
          value: e.value,
          target: l.current,
          syntheticEvent: t
        }
      );
    },
    [e.onClick, e.value]
  ), r = n.useCallback(
    (t) => {
      e.onKeyDown && e.onKeyDown.call(
        void 0,
        {
          target: l.current,
          syntheticEvent: t
        }
      );
    },
    [e.onKeyDown]
  ), d = n.useCallback(
    (t) => {
      e.onFocus && e.onFocus.call(
        void 0,
        {
          target: l.current,
          syntheticEvent: t
        }
      );
    },
    [e.onFocus]
  ), v = n.useCallback(
    (t) => {
      e.onBlur && e.onBlur.call(
        void 0,
        {
          target: l.current,
          syntheticEvent: t
        }
      );
    },
    [e.onBlur]
  ), f = n.useCallback(
    (t) => {
      e.onMouseMove && e.onMouseMove.call(
        void 0,
        {
          target: l.current,
          syntheticEvent: t
        }
      );
    },
    [e.onMouseMove]
  ), h = n.useCallback(
    (t) => {
      e.onMouseEnter && e.onMouseEnter.call(
        void 0,
        {
          target: l.current,
          syntheticEvent: t
        }
      );
    },
    [e.onMouseEnter]
  ), M = n.useCallback(
    (t) => {
      e.onMouseLeave && e.onMouseLeave.call(
        void 0,
        {
          target: l.current,
          syntheticEvent: t
        }
      );
    },
    [e.onMouseLeave]
  );
  return /* @__PURE__ */ n.createElement(
    "span",
    {
      id: e.id,
      ref: o,
      dir: u,
      "data-half": e.half,
      tabIndex: e.tabIndex,
      title: e.title,
      style: e.style,
      className: y(
        "k-rating-item",
        {
          "k-rtl": u === "rtl",
          "k-selected": e.selected,
          "k-hover": e.hovered
        },
        e.className
      ),
      onClick: s,
      onKeyDown: r,
      onFocus: d,
      onBlur: v,
      onMouseEnter: h,
      onMouseLeave: M,
      onMouseMove: f
    },
    e.children
  );
}), k = {
  id: a.string,
  icon: a.any,
  dir: a.oneOf(["ltr", "rtl"]),
  title: a.string,
  value: a.number,
  half: a.bool,
  selected: a.bool,
  hovered: a.bool
}, b = {
  dir: "ltr"
};
c.displayName = "KendoReactRatingItem";
c.propTypes = k;
c.defaultProps = b;
export {
  c as RatingItem
};
