// @progress/kendo-react-inputs/rating/Rating.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as a from "react";
import i from "prop-types";
import { createPropsContext as ee, usePropsContext as te, useDir as ae, Keys as M, IconWrap as f, toIconName as B, getTabIndex as ne, classNames as le, svgIconPropType as $ } from "@progress/kendo-react-common";
import { starOutlineIcon as K, starIcon as L } from "@progress/kendo-svg-icons";
import { RatingItem as oe } from "./RatingItem.mjs";
import { RATING_ACTION as r, ratingReducer as se } from "./rating-reducer.mjs";
import { calcIsFirstHalf as A, toRound as R, getRemainder as ie, isHalf as re, isSelected as F, isCorrectValue as X } from "./utils/index.mjs";
import { ratingAriaLabel as q, messages as ce } from "../messages/index.mjs";
import { useLocalization as ue } from "@progress/kendo-react-intl";
const G = (n, g, e) => {
  const [h, c] = a.useState(n);
  return [h, (o) => {
    const V = se(g.state || h, { ...o, ...g });
    e && e(V, o.event), c(V);
  }];
}, me = ee(), H = a.forwardRef((n, g) => {
  const e = te(me, n), h = ue(), c = a.useRef(null), D = a.useRef(null), o = ae(D, e.dir);
  a.useImperativeHandle(c, () => ({
    element: D.current,
    props: e
  })), a.useImperativeHandle(g, () => c.current);
  const V = a.useCallback(
    (t, l) => {
      e.onChange && c.current && e.onChange.call(void 0, {
        value: t,
        target: c.current,
        syntheticEvent: l
      });
    },
    [e.onChange]
  ), b = a.useMemo(
    () => e.min !== void 0 ? e.min : m.min,
    [e.min]
  ), I = a.useMemo(
    () => e.max !== void 0 ? e.max : m.max,
    [e.max]
  ), d = a.useMemo(
    () => e.precision || m.precision,
    [e.precision]
  ), s = a.useMemo(
    () => e.step !== void 0 ? e.step : m.step,
    [e.step]
  ), w = a.useMemo(
    () => e.readonly || m.readonly,
    [e.readonly]
  ), k = a.useMemo(
    () => e.disabled || m.disabled,
    [e.disabled]
  ), W = a.useMemo(
    () => e.item || m.item,
    [e.item]
  ), [P, p] = G(
    e.defaultValue !== void 0 ? e.defaultValue : m.defaultValue,
    {
      state: e.value,
      min: b,
      max: I,
      step: d === "half" ? s / 2 : s
    },
    V
  ), [T, O] = G(
    null,
    {
      state: e.value,
      min: b,
      max: I,
      step: d === "half" ? s / 2 : s,
      precision: d
    }
  ), u = a.useMemo(
    () => e.value !== void 0 ? e.value : P,
    [e.value, P]
  ), x = a.useMemo(
    () => T,
    [T]
  ), z = a.useMemo(
    () => e.selection || m.selection,
    [e.selection]
  ), C = a.useMemo(
    () => s / (d === "half" ? 2 : 1),
    [s, d]
  ), _ = a.useCallback(
    (t) => {
      if (!(!t.target || !t.target.element || w || k)) {
        if (d === "half") {
          const l = t.target.element.getBoundingClientRect(), E = A(o || "ltr", l, t.syntheticEvent.clientX) ? R(t.target.value - s / 2, C) : t.target.value;
          p({ type: r.select, payload: E, event: t.syntheticEvent });
        } else
          p({ type: r.select, payload: t.target.value, event: t.syntheticEvent });
        if (e.onClick) {
          const l = {
            value: u,
            target: c.current,
            syntheticEvent: t.syntheticEvent
          };
          e.onClick.call(void 0, l);
        }
      }
    },
    [e.onClick, p, u]
  ), j = a.useCallback(
    (t) => {
      if (!(w || k)) {
        switch (t.keyCode) {
          case M.right:
            t.preventDefault(), p({
              type: o === "rtl" ? r.decrease : r.increase,
              event: t
            });
            break;
          case M.left:
            t.preventDefault(), p({
              type: o === "rtl" ? r.increase : r.decrease,
              event: t
            });
            break;
          case M.home:
            t.preventDefault(), p({
              type: o === "rtl" ? r.min : r.max,
              event: t
            });
            break;
          case M.end:
            t.preventDefault(), p({
              type: o === "rtl" ? r.max : r.min,
              event: t
            });
            break;
          case M.esc:
            t.preventDefault(), p({
              type: r.deselect,
              event: t
            });
            break;
        }
        e.onKeyDown && e.onKeyDown.call(
          void 0,
          {
            value: u,
            target: c.current,
            syntheticEvent: t
          }
        );
      }
    },
    [e.onKeyDown, p, u]
  ), J = a.useCallback(
    (t) => {
      if (e.onFocus) {
        const l = {
          target: c.current,
          syntheticEvent: t
        };
        e.onFocus.call(
          void 0,
          l
        );
      }
    },
    [e.onFocus]
  ), Q = a.useCallback(
    (t) => {
      if (e.onBlur) {
        const l = {
          target: c.current,
          syntheticEvent: t
        };
        e.onBlur.call(
          void 0,
          l
        );
      }
    },
    [e.onBlur]
  ), U = a.useCallback(
    (t) => {
      if (!(!t.target || !t.target.element))
        if (d === "half") {
          const l = t.target.element.getBoundingClientRect(), E = A(o || "ltr", l, t.syntheticEvent.clientX) ? t.target.value - s / 2 : t.target.value;
          O({ type: r.select, payload: E, event: t.syntheticEvent });
        } else
          O({ type: r.select, payload: t.target.value, event: t.syntheticEvent });
    },
    [d, s, o]
  ), Y = a.useCallback(
    (t) => {
      O({ type: r.reset, event: t.syntheticEvent });
    },
    []
  ), S = [], Z = ie(R(I - b, C), s);
  for (let t = b; t <= I; t = R(t + s, C)) {
    const l = R(t + Z, C), v = d === "half" ? re(l, x !== null ? x : u !== null ? u : 0, s) : !1, E = F(l, u, s, z), N = F(l, x !== null ? x : u, s, z), y = F(l, x, s, z);
    S.push(
      /* @__PURE__ */ a.createElement(
        W,
        {
          key: l,
          value: l,
          dir: o,
          title: String(v ? R(l - s / 2, C) : l),
          icon: e.icon,
          svgIcon: e.svgIcon,
          half: v,
          selected: E || N,
          hovered: y,
          onClick: _,
          onMouseMove: U,
          onMouseLeave: Y
        },
        v && /* @__PURE__ */ a.createElement("span", { className: "k-rating-precision-complement" }, !(e.icon || e.svgIconOutline) && /* @__PURE__ */ a.createElement(f, { name: "star-outline", icon: K, style: { clipPath: `${o === "rtl" ? "inset(0 50% 0 0)" : "inset(0 0 0 50%)"}` }, size: "xlarge" }), (e.icon || e.svgIconOutline) && /* @__PURE__ */ a.createElement(f, { name: B(`${e.icon}-outline`), icon: e.svgIconOutline, size: "xlarge" })),
        v && /* @__PURE__ */ a.createElement("span", { className: "k-rating-precision-part" }, !(e.icon || e.svgIcon) && /* @__PURE__ */ a.createElement(f, { name: "star", icon: L, style: { clipPath: `${o === "rtl" ? "inset(0 0 0 50%)" : "inset(0 50% 0 0)"}` }, size: "xlarge" }), (e.icon || e.svgIcon) && /* @__PURE__ */ a.createElement(f, { name: e.icon ? B(e.icon) : void 0, icon: e.svgIcon, size: "xlarge" })),
        v && /* @__PURE__ */ a.createElement("span", { style: { width: "24px", height: "24px", display: "block" } }),
        !v && !(e.icon || e.svgIcon) && (y || N && !y) && /* @__PURE__ */ a.createElement(f, { name: "star", icon: L, size: "xlarge" }) || !(e.icon || e.svgIcon) && !y && !N && /* @__PURE__ */ a.createElement(f, { name: "star-outline", icon: K, size: "xlarge" }) || (e.icon || e.svgIcon) && (y || N && !y) && /* @__PURE__ */ a.createElement(f, { name: e.icon, icon: e.svgIcon, size: "xlarge" }) || (e.icon || e.svgIcon) && !y && /* @__PURE__ */ a.createElement(f, { name: B(`${e.icon}-outline`), icon: e.svgIconOutline, size: "xlarge" })
      )
    );
  }
  return /* @__PURE__ */ a.createElement(
    "span",
    {
      id: e.id,
      style: e.style,
      ref: D,
      role: "slider",
      dir: o,
      tabIndex: ne(e.tabIndex, e.disabled, void 0),
      className: le(
        "k-rating",
        {
          "k-rtl": o === "rtl",
          "k-readonly": w,
          "k-disabled": k
        },
        e.className
      ),
      onKeyDown: j,
      onFocus: J,
      onBlur: Q,
      "aria-valuemin": b,
      "aria-valuemax": I,
      "aria-valuenow": u !== null ? u : void 0,
      "aria-disabled": k ? "true" : void 0,
      "aria-label": h.toLanguageString(q, ce[q]),
      "aria-labelledby": e.ariaLabelledBy,
      "aria-describedby": e.ariaDescribedBy
    },
    /* @__PURE__ */ a.createElement(
      "input",
      {
        id: "rating",
        className: "k-hidden",
        readOnly: w,
        disabled: k
      }
    ),
    /* @__PURE__ */ a.createElement(
      "span",
      {
        className: "k-rating-container"
      },
      S
    ),
    e.label && /* @__PURE__ */ a.createElement(
      "span",
      {
        className: "k-rating-label"
      },
      e.label
    )
  );
}), de = {
  id: i.string,
  dir: i.oneOf(["ltr", "rtl"]),
  selection: i.oneOf(["continues", "single"]),
  precision: i.oneOf(["item", "half"]),
  value: (n, g, e) => n.value && n.min && n.max && n.step && !(n.precision === "half" ? X(n.min, n.max, n.step / 2, n.value) : X(n.min, n.max, n.step, n.value)) && n.precision === "half" ? new Error(
    `Invalid prop + ${g} supplied to ${e}.
                    The value of the { value } property is not correct, please check your values.
                    `
  ) : null,
  defaultValue: i.number,
  min: i.number,
  max: i.number,
  step: (n, g, e) => n.step !== void 0 && n.step <= 0 ? new Error(
    `Invalid prop + ${g} supplied to ${e}.
                The value of the { step } property is cannot be equal or less than "0", please check your values.
                `
  ) : null,
  hovered: i.number,
  label: i.string,
  readonly: i.bool,
  disabled: i.bool,
  half: i.bool,
  icon: i.string,
  svgIcon: $,
  svgIconOutline: $,
  ariaDescribedBy: i.string
}, m = {
  dir: "ltr",
  min: 1,
  max: 5,
  step: 1,
  item: oe,
  defaultValue: null,
  readonly: !1,
  disabled: !1,
  selection: "continues",
  precision: "item"
};
H.displayName = "KendoReactRating";
H.propTypes = de;
H.defaultProps = m;
export {
  H as Rating,
  me as RatingPropsContext
};
