// @progress/kendo-react-inputs/colors/FlatColorPicker.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import l from "prop-types";
import { validatePackage as q, getTabIndex as D, classNames as K } from "@progress/kendo-react-common";
import { ButtonGroup as _, Button as u } from "@progress/kendo-react-buttons";
import { dropletSliderIcon as j, paletteIcon as J, dropletSlashIcon as O } from "@progress/kendo-svg-icons";
import { ColorPalette as Q } from "./ColorPalette.mjs";
import { ColorGradient as U } from "./ColorGradient.mjs";
import { packageMetadata as W } from "../package-metadata.mjs";
import { useLocalization as X } from "@progress/kendo-react-intl";
import { flatColorPickerColorGradientBtn as w, messages as v, flatColorPickerColorPaletteBtn as B, flatColorPickerClearBtn as N, flatColorPickerCancelBtn as Y, flatColorPickerApplyBtn as Z } from "../messages/index.mjs";
const I = "rgba(255, 255, 255, 1)", L = e.forwardRef((a, G) => {
  q(W);
  const s = e.useRef(null), i = e.useRef(null), S = e.useRef(null), { defaultValue: r = I, showButtons: C = !0, showPreview: x = !0, showClearButton: R = !0 } = a, [c, F] = e.useState((a.view || "ColorGradient") === "ColorGradient"), [o, m] = e.useState(a.value || r), [d, k] = e.useState(a.value || r), g = a.value !== void 0 ? a.value : d;
  e.useEffect(() => {
    m(a.value || r);
  }, [a.value, r]);
  const f = X(), b = e.useCallback(
    () => {
      s.current && s.current.focus();
    },
    [s]
  );
  e.useImperativeHandle(s, () => ({
    element: i.current,
    focus: b,
    props: a,
    value: g
  }), [g, b, a]), e.useImperativeHandle(G, () => s.current);
  const h = e.useCallback(
    (t) => {
      (t === "ColorGradient" && !c || t === "ColorPalette" && c) && F(!c);
    },
    [c]
  ), V = e.useCallback(
    (t) => {
      if (m(r), a.onPreviewChange) {
        const n = {
          value: r,
          nativeEvent: t.nativeEvent,
          syntheticEvent: t
        };
        a.onPreviewChange.call(void 0, n);
      }
    },
    [r, a.onPreviewChange]
  ), E = e.useCallback(
    (t) => {
      if (m(t.value), a.onPreviewChange) {
        const n = {
          value: t.value,
          nativeEvent: t.nativeEvent,
          syntheticEvent: t.syntheticEvent
        };
        a.onPreviewChange.call(void 0, n);
      }
    },
    [a.onPreviewChange]
  ), T = e.useCallback(
    (t) => {
      if (k(o), a.onChange) {
        const n = {
          value: o,
          nativeEvent: t.nativeEvent,
          syntheticEvent: t
        };
        a.onChange.call(void 0, n);
      }
    },
    [o, a.onChange]
  ), M = e.useCallback(
    () => {
      k(I);
    },
    []
  ), H = e.useCallback(
    (t) => {
      if (m(d), a.onPreviewChange) {
        const n = {
          value: d,
          nativeEvent: t.nativeEvent,
          syntheticEvent: t
        };
        a.onPreviewChange.call(void 0, n);
      }
    },
    [d, a.onPreviewChange]
  ), y = e.useCallback(
    (t) => {
      i.current && !(t.nativeEvent.target instanceof HTMLInputElement) && i.current.focus();
    },
    [i]
  ), z = e.useCallback(
    (t) => {
      var P;
      if ((!t.relatedTarget || !((P = i.current) != null && P.contains(t.relatedTarget))) && (C || k(o), !C && a.onChange)) {
        const A = {
          value: o,
          nativeEvent: t.nativeEvent,
          syntheticEvent: t
        };
        a.onChange.call(void 0, A);
      }
    },
    [o, C, a.onChange]
  );
  return /* @__PURE__ */ e.createElement(
    "div",
    {
      id: a.id,
      role: "textbox",
      "aria-label": a.ariaLabel,
      "aria-labelledby": a.ariaLabelledBy,
      "aria-disabled": a.disabled ? "true" : void 0,
      style: a.style,
      ref: i,
      tabIndex: D(a.tabIndex, a.disabled),
      className: K(
        "k-flatcolorpicker k-coloreditor",
        {
          "k-disabled": a.disabled
        },
        a.className
      ),
      onBlur: z
    },
    /* @__PURE__ */ e.createElement(e.Fragment, null, a.header || /* @__PURE__ */ e.createElement("div", { className: "k-coloreditor-header k-hstack" }, /* @__PURE__ */ e.createElement("div", { className: "k-coloreditor-header-actions k-hstack" }, /* @__PURE__ */ e.createElement(_, { className: "k-button-group-flat" }, /* @__PURE__ */ e.createElement(
      u,
      {
        type: "button",
        "aria-label": f.toLanguageString(
          w,
          v[w]
        ),
        togglable: !0,
        fillMode: "flat",
        selected: c,
        onClick: () => h("ColorGradient"),
        icon: "droplet-slider",
        svgIcon: j
      }
    ), /* @__PURE__ */ e.createElement(
      u,
      {
        type: "button",
        "aria-label": f.toLanguageString(
          B,
          v[B]
        ),
        togglable: !0,
        fillMode: "flat",
        selected: !c,
        onClick: () => h("ColorPalette"),
        icon: "palette",
        svgIcon: J
      }
    ))), /* @__PURE__ */ e.createElement("div", { className: "k-spacer" }), /* @__PURE__ */ e.createElement("div", { className: "k-coloreditor-header-actions k-hstack" }, R && /* @__PURE__ */ e.createElement(
      u,
      {
        type: "button",
        fillMode: "flat",
        onClick: V,
        "aria-label": f.toLanguageString(
          N,
          v[N]
        ),
        icon: "droplet-slash",
        svgIcon: O
      }
    ), x && /* @__PURE__ */ e.createElement("div", { className: "k-coloreditor-preview k-vstack" }, /* @__PURE__ */ e.createElement("span", { className: "k-coloreditor-preview-color k-color-preview", style: { background: o } }), /* @__PURE__ */ e.createElement("span", { className: "k-coloreditor-current-color k-color-preview", style: { background: d }, onClick: H })))), /* @__PURE__ */ e.createElement("div", { className: "k-coloreditor-views k-vstack" }, c ? /* @__PURE__ */ e.createElement(
      U,
      {
        ref: S,
        role: "none",
        tabIndex: -1,
        ariaLabel: void 0,
        value: o,
        onChange: E,
        onFocus: y,
        opacity: a.opacity,
        format: a.format
      }
    ) : /* @__PURE__ */ e.createElement(Q, { ariaDisabled: !0, ariaLabelledBy: "required_label", value: o, onChange: E, onFocus: y })), C && /* @__PURE__ */ e.createElement("div", { className: "k-coloreditor-footer k-actions k-actions-end" }, /* @__PURE__ */ e.createElement(u, { type: "button", className: "k-coloreditor-cancel", onClick: M }, v[Y]), /* @__PURE__ */ e.createElement(u, { type: "button", className: "k-coloreditor-apply k-primary", onClick: T }, v[Z])))
  );
}), $ = {
  id: l.string,
  style: l.any,
  className: l.string,
  value: l.string,
  defaultValue: l.string,
  onPreviewChange: l.func,
  onChange: l.func,
  opacity: l.bool,
  format: l.any,
  tabIndex: l.number,
  ariaLabel: l.string,
  ariaLabelledBy: l.string,
  disabled: l.bool,
  view: l.any,
  header: l.any,
  showClearButton: l.bool,
  showPreview: l.bool,
  showButtons: l.bool
};
L.displayName = "KendoFlatColorPicker";
L.propTypes = $;
export {
  L as FlatColorPicker
};
