// @progress/kendo-react-grid/header/HeaderRow.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as r from "react";
import { normalize as G } from "../interfaces/GridSortSettings.mjs";
import { ColumnDraggable as L } from "../drag/ColumnDraggable.mjs";
import { GridHeaderCell as O } from "./GridHeaderCell.mjs";
import { classNames as P, Keys as u, IconWrap as V } from "@progress/kendo-react-common";
import { ColumnResizer as W, HeaderThElement as q } from "@progress/kendo-react-data-tools";
import { useLocalization as B } from "@progress/kendo-react-intl";
import { sortAriaLabel as z, messages as T } from "../messages/index.mjs";
import { sortAscSmallIcon as U, sortDescSmallIcon as j } from "@progress/kendo-svg-icons";
import { resolveCells as J } from "../utils/index.mjs";
const Q = {
  true: { asc: "desc", desc: "", "": "asc" },
  false: { asc: "desc", desc: "asc", "": "asc" }
}, X = {
  none: "none",
  asc: "ascending",
  desc: "descending"
}, ne = (e) => {
  const [H, b] = r.useState(), D = B(), f = r.useRef(null);
  let E = 0, M = -1;
  const k = (t, a) => {
    if (t.preventDefault(), !e.sortChange)
      return;
    const { allowUnsort: l, mode: c } = G(e.sortable || !1, a.sortable || !1), o = (e.sort || []).filter((i) => i.field === a.field)[0], d = Q[l][o && o.dir || ""], s = c === "single" ? [] : (e.sort || []).filter((i) => i.field !== a.field);
    d !== "" && a.field && s.push({ field: a.field, dir: d }), e.sortChange(s, t);
  }, x = (t, a) => {
    if (t.isDefaultPrevented())
      return;
    const { keyCode: l, metaKey: c, ctrlKey: o, altKey: d } = t, s = e.sortable && a.sortable, i = t.target;
    if (s && t.keyCode === u.enter && i.className.indexOf("k-table-th") !== -1 && k(t, a), e.navigatable) {
      if (e.groupable && (c && l === u.enter || o && l === u.space)) {
        t.preventDefault();
        const m = e.columns.findIndex((g) => g.field === a.field);
        e.columnGroupChange && e.columnGroupChange(m, t);
      }
      d && l === u.down && (t.preventDefault(), b(a.field), f.current = i);
    }
  }, I = () => {
    b(void 0), f.current && f.current.getAttribute("tabindex") !== "-1" && f.current.focus();
  }, v = (t) => e.sort ? t >= 0 && [
    /* @__PURE__ */ r.createElement("span", { key: 1, className: "k-sort-icon" }, /* @__PURE__ */ r.createElement(
      V,
      {
        name: "sort-" + e.sort[t].dir + "-small",
        icon: e.sort[t].dir === "asc" ? U : j
      }
    )),
    e.sort.length > 1 && /* @__PURE__ */ r.createElement("span", { key: 2, className: "k-sort-icon" }, /* @__PURE__ */ r.createElement("span", { className: "k-sort-order" }, t + 1))
  ] : null, y = (t) => t.map((a) => {
    const l = e.columns[a], c = e.sortable && l.sortable, o = e.sort ? e.sort.findIndex((n) => n.field === l.field) : -1, d = o >= 0 && e.sort[o].dir || "none", s = l.columnMenu === null ? null : l.columnMenu || e.columnMenu, i = P({
      "k-first": l.kFirst,
      "k-filterable": !!s,
      "k-table-th": !0,
      "k-header": !0,
      "k-grid-header-sticky": l.locked,
      "k-sorted": e.sort && e.sort.some((n) => n.field === l.field)
    }, l.headerClassName);
    l.locked === !1 && (l.left = 0);
    const m = l.left !== void 0 ? e.isRtl ? { left: l.right, right: l.left } : { left: l.left, right: l.right } : {}, g = D.toLanguageString(z, T[z]), N = l.isAccessible ? {
      ariaSort: X[d],
      role: "columnheader",
      ariaColumnIndex: l.ariaColumnIndex,
      ariaSelected: !1,
      ariaDescription: c ? g : ""
    } : {
      role: "presentation"
    }, h = l.declarationIndex >= 0 ? ++M : --E, K = l.headerCell ? l.headerCell : O, R = /* @__PURE__ */ r.createElement(
      K,
      {
        key: 1,
        field: l.field,
        onClick: c && ((n) => k(n, l)) || void 0,
        selectionChange: e.selectionChange,
        title: l.title,
        selectionValue: l.headerSelectionValue,
        render: e.cellRender,
        children: v(o),
        columnMenuWrapperProps: {
          column: {
            field: l.field,
            title: l.title,
            locked: l.locked,
            filter: l.filter,
            id: l.id
          },
          sortable: c && e.sortable,
          sort: e.sort,
          onSortChange: e.sortChange,
          filter: e.filter,
          filterable: e.filterable && l.filterable,
          filterOperators: e.filterOperators,
          onFilterChange: e.filterChange,
          show: H === l.field ? !0 : void 0,
          onCloseMenu: I,
          navigatable: e.navigatable,
          group: e.group,
          groupable: e.groupable,
          onGroupChange: e.groupChange,
          columnMenu: s
        }
      }
    ), S = {
      ...N,
      key: h,
      colSpan: l.colSpan,
      rowSpan: l.rowSpan,
      className: i,
      style: m,
      columnId: l.id,
      navigatable: l.navigatable,
      onKeyDown: (n) => x(n, l),
      role: "columnheader"
    }, w = [
      R,
      e.columnResize && e.columnResize.resizable && l.resizable && /* @__PURE__ */ r.createElement(
        W,
        {
          key: 2,
          resize: (n, A, F) => e.columnResize && e.columnResize.dragHandler(n, l, A, F),
          autofit: (n) => e.columnResize && e.columnResize.dblClickHandler(n, [l.id])
        }
      )
    ], C = J(e.cells, l.cells);
    if (C && C.headerCell) {
      const n = C.headerCell;
      return /* @__PURE__ */ r.createElement(n, { ...R.props, key: h, thProps: S, index: a }, w);
    }
    return /* @__PURE__ */ r.createElement(
      q,
      {
        ...S,
        key: h
      },
      w
    );
  });
  return e.columnsMap.map((t, a) => e.pressHandler && /* @__PURE__ */ r.createElement(
    L,
    {
      key: a,
      pressHandler: e.pressHandler,
      dragHandler: e.dragHandler,
      releaseHandler: e.releaseHandler,
      ariaRowIndex: a + 1,
      dragClue: e.dragClue,
      headerRef: e.headerRef,
      containerRef: e.containerRef
    },
    y(t)
  ) || /* @__PURE__ */ r.createElement("tr", { key: a, className: "k-table-row", role: "row", "aria-rowindex": a + 1 }, y(t)));
};
export {
  ne as HeaderRow
};
