// @progress/kendo-react-upload/UploadList.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as s from "react";
import { classNames as n } from "@progress/kendo-react-common";
import { UploadListGroup as i } from "./UploadListGroup.mjs";
class u extends s.Component {
  mapListGroups() {
    const { groupedFiles: e, navigationIndex: r, ...o } = this.props;
    return Object.keys(e).map((t, a) => {
      const p = e[t];
      return /* @__PURE__ */ s.createElement(
        i,
        {
          key: t,
          files: p,
          index: a,
          navigationIndex: r,
          ...o
        }
      );
    });
  }
  render() {
    const e = n(
      "k-upload-files",
      "k-reset"
    );
    return /* @__PURE__ */ s.createElement("ul", { className: e }, this.mapListGroups());
  }
}
export {
  u as UploadList
};
